### PR TITLE
feat(graph/privesc): pathfinding.cloud gap analysis — 47 new AWS privesc paths + Concern B fix (LAB-3908)

### PR DESCRIPTION
**Pull request by Elgin-Lee**
## Summary

Reconciles Aurelian's AWS privilege-escalation graph coverage against all 122 techniques catalogued at [pathfinding.cloud/labs](https://pathfinding.cloud/labs/). Adds 47 new Cypher enrichment queries (methods 43–89), expands the IAM action allowlist, adds Concern B fixes, and ships a comprehensive three-tier test suite.

**New services covered:** App Runner, AWS Batch, Amazon Braket, CloudFormation StackSets, CodeDeploy, Cognito Identity, EC2 Instance Connect, EC2 instance-profile swap, ECS CreateService/StartTask/ExecuteCommand, EMR RunJobFlow, EMR Serverless, GameLift, Glue CreateDevEndpoint/UpdateJob/CreateSession/CreateJob+Trigger/UpdateJob+Trigger, EC2 Image Builder, Kinesis Analytics v1 (create+start), Lambda UpdateCode+AddPermission, Lambda UpdateCode+InvokeFunction, HealthOmics, SageMaker lifecycle config, EventBridge Scheduler, SSM CreateDocument+StartAutomation, Step Functions (create + update, both with and without StartExecution), Bedrock AgentCore, EC2 RequestSpotInstances, EC2 CreateLaunchTemplateVersion, EC2 ModifyInstanceAttribute, Amplify (CreateApp+CreateBranch+StartJob), and compound execution-gated variants (Glue StartJobRun, EMR Serverless StartJobRun, Kinesis Analytics StartApplication, HealthOmics StartRun, GameLift CreateBuild+CreateFleet, Image Builder CreateImage).

**Coverage summary** (against 122 pathfinding.cloud labs):

| Status | Count |
|--------|-------|
| Covered by existing methods 01–42 | 45 labs |
| Partial (individual perms detected, compound path not single-edge) | 6 labs |
| Implemented in this PR (methods 43–89) | 47 new methods |
| Out of scope (bucket-target, cross-account multi-hop, CSPM) | ~24 labs |

Full classification in `docs/gap_analysis_pathfinding_cloud_privesc.md`.

---

## Key fixes in this PR
PassRole+service methods were creating `CAN_PRIVESC` edges to synthetic service resource nodes (e.g. `arn:aws:apprunner:*:*:*`) instead of `:Principal` nodes. The `aws/analysis/privesc_paths` query only traverses edges between `:Principal` nodes, so all 47 new methods produced zero findings in the analysis output.

**Fix:** All 47 new PassRole+service methods scope `victim` to the specific passed IAM role using `MATCH (attacker:Principal)-[passrole]->(victim:Principal) WHERE type(passrole) = 'IAM_PASSROLE' WITH attacker, victim ... MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)`. IAM roles are labeled `:Principal` in the graph, so this is both precise and correct. Pre-existing methods 14–42 have the same issue, tracked separately in [LAB-3947](https://linear.app/praetorianlabs/issue/LAB-3947).

Lambda-only methods (39, 65) without PassRole use `(target)-[:HAS_ROLE]->(victim:Principal)` to reach the Lambda execution role via the existing `resource_to_role` enrichment.

### Tier 1 — Neo4j container (89 methods, all passing)

| Test | What it proves |
|------|---------------|
| `TestPrivescQueriesNeo4j` | 89 per-method TP cases — each seeds the exact required permissions and asserts ≥1 `CAN_PRIVESC` edge to the correct target |
| `TestPrivescNegativePermissions` | 7 FP guard cases: PassRole-only, service-only, compound missing one action, PassRole to `:Resource` (not `:Principal`), 3-action compound missing one action, SSM cross-resource |
| `TestPrivescEdgeMetadata` | Asserts `r.method` and `r.severity` are set correctly on `CAN_PRIVESC` edges (wrong value would pass count assertions but break pathfinding FP tests) |
| `TestPrivescEnrichAWSIdempotent` | Running `EnrichAWS` twice produces the same edge count — guards against `CREATE` instead of `MERGE` regressions |
| `TestPrivescAnalysisQuery` | 1/2/3-hop paths found, >3-hop excluded, admin-as-attacker excluded, no self-loops |
| `TestPassRoleServiceFanOutReachesAnalysisQuery` | End-to-end: PassRole+AppRunner seeds → `EnrichAWS` → `aws/analysis/privesc_paths` finds the attacker |
| `TestPrivescMultiHopThroughPassRoleMethod` | 2-hop chain: attacker → intermediate (via method_43 scoped edge) → admin (via method_01 fan-out) |
| `TestPrivescMultiHopPaths` | 4-principal graph; standalone methods create principal-to-principal chains detectable by analysis |

### Tier 2 — Real AWS E2E (Terraform fixture in account 741011675301)

3 fixture IAM users, passable role with `AdministratorAccess`, and deployed service resources (ECS cluster, Step Functions state machine, Glue job, Batch queue, Cognito pool, SageMaker lifecycle, Image Builder config). Assertions added beyond original floor counts:

- `passrole_user_targets_passable_role`: `CAN_PRIVESC` from PassRole user must point to the specific passable role ARN
- `passrole_user_edge_count_not_fanout`: total `CAN_PRIVESC` count ≤ 10 — detects fan-out regression (previously produced 136+ edges to wrong principals)
- `passrole_user_appears_in_analysis_output`: runs `aws/analysis/privesc_paths` and asserts the attacker appears, proving end-to-end traversability

### Tier 3 — pathfinding.cloud fixture (103 IAM users, requires AWS)

151 table-driven test cases: 109 TP (each IAM user has the specific permissions for one technique; asserts ≥1 `CAN_PRIVESC` edge) and 42 FP (asserts `r.method` never matches the specific method suffix for principals that should NOT trigger it).

---

## Notes

- Pre-existing methods 14–42 have the same Concern B architecture issue (edge points to service resource instead of `:Principal`). Tracked in [LAB-3947](https://linear.app/praetorianlabs/issue/LAB-3947) — out of scope here
- The 12 non-PassRole, non-Lambda fan-out methods (44, 46, 50, 52, 53, 56, 67, 71, 72, 74, 76) still use the fan-out pattern. Their victim is a service resource's associated IAM role (e.g. ECS task role, Batch job role) that would require additional `HAS_ROLE`-style enrichment per service — deferred as a follow-up.
- Azure/GCP privesc paths deferred until IAM graphs land (LAB-3742).

### CODE REVIEW

File: pkg/graph/queries/enrich/aws/privesc/method_65_lambda_add_permission.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  WHERE type(add_perm) = 'LAMBDA_ADDPERMISSION'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
+  ON CREATE SET pe._created = timestamp()
+  SET pe.method = 'lambda:UpdateFunctionCode + lambda:AddPermission',
+      pe.severity = 'high',
+      pe._updated = timestamp()
+  RETURN count(pe) as edges_created
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_65_lambda_add_permission.yaml]: Fixed. Changed fan-out to traverse the Lambda function's execution role via `(target)-[:HAS_ROLE]->(victim:Principal)`, using the existing `resource_to_role` enrichment that links Lambda functions to their IAM roles. Also updated the matching test case to seed a `HAS_ROLE` edge and verify the edge targets the execution role.

File: test/terraform/aws/recon/privesc/main.tf [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+  name = "${local.prefix}-passable-role"
+
+  assume_role_policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [{
+      Effect    = "Allow"
+      Action    = "sts:AssumeRole"
+      Principal = { Service = "lambda.amazonaws.com" }
+    }]
+  })
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply test/terraform/aws/recon/privesc/main.tf:61]: Fixed. Broadened the passable role's trust policy to include all 15 service principals used by the fixture methods: `ecs-tasks.amazonaws.com`, `states.amazonaws.com`, `glue.amazonaws.com`, `ec2.amazonaws.com`, `batch.amazonaws.com`, `apprunner.amazonaws.com`, `imagebuilder.amazonaws.com`, `sagemaker.amazonaws.com`, `braket.amazonaws.com`, `emr-serverless.amazonaws.com`, `scheduler.amazonaws.com`, `ssm.amazonaws.com`, `omics.amazonaws.com`, `kinesisanalytics.amazonaws.com`.

File: pkg/graph/queries/enrich/aws/privesc/method_38_autoscaling_launch_template.yaml [hunk 3]
```diff
@@ -18,7 +18,10 @@ cypher: |
   WITH attacker, service_resource
   MATCH (attacker)-[asg]->(asg_resource)
   WHERE type(asg) = 'AUTOSCALING_CREATEAUTOSCALINGGROUP'
-  MERGE (attacker)-[pe:CAN_PRIVESC]->(service_resource)
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_38_autoscaling_launch_template.yaml]: Fixed. Changed to `MATCH (attacker:Principal)-[passrole]->(victim:Principal) WHERE type(passrole) = 'IAM_PASSROLE' WITH attacker, victim` so the victim is scoped to the specific passed IAM role rather than all principals. IAM roles are labeled `:Principal` in the graph (confirmed in `pkg/graph/transformers/aws/aws.go:51`), so this pattern is both safe and eliminates false positives. Validated by Codex before applying.

File: pkg/graph/queries/enrich/aws/privesc/method_40_bedrock_create_code_interpreter.yaml [hunk 4]
```diff
@@ -14,8 +14,11 @@ cypher: |
   WHERE type(passrole) = 'IAM_PASSROLE'
   WITH attacker
   MATCH (attacker)-[create]->(service_resource)
-  WHERE type(create) = 'BEDROCK_AGENTCORE_CREATECODEINTERPRETER'
-  MERGE (attacker)-[pe:CAN_PRIVESC]->(service_resource)
+  WHERE type(create) = 'BEDROCK-AGENTCORE_CREATECODEINTERPRETER'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_40_bedrock_create_code_interpreter.yaml]: Fixed (fan-out removed). Also corrected the description which incorrectly claimed 3 permissions were required — the Cypher only requires `iam:PassRole + bedrock-agentcore:CreateCodeInterpreter`. Description updated to match.

File: pkg/graph/queries/enrich/aws/privesc/method_47_braket_create_job.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+cypher: |
+  MATCH (attacker:Principal)-[passrole]->(target)
+  WHERE type(passrole) = 'IAM_PASSROLE'
+  WITH attacker
+  MATCH (attacker)-[create]->(service_resource)
+  WHERE type(create) = 'BRAKET_CREATEJOB'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_47_braket_create_job.yaml]: Fixed. Scoped `CAN_PRIVESC` to the specific passed IAM role using `MATCH (attacker:Principal)-[passrole]->(victim:Principal) WHERE type(passrole) = 'IAM_PASSROLE' WITH attacker, victim ... WITH DISTINCT attacker, victim MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)`.

File: pkg/graph/queries/enrich/aws/privesc/method_48_cloudformation_create_stackset.yaml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+cypher: |
+  MATCH (attacker:Principal)-[passrole]->(target)
+  WHERE type(passrole) = 'IAM_PASSROLE'
+  WITH attacker
+  MATCH (attacker)-[create]->(service_resource)
+  WHERE type(create) = 'CLOUDFORMATION_CREATESTACKSET'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_48_cloudformation_create_stackset.yaml]: Fixed. Same scoped-victim pattern applied.

File: pkg/graph/queries/enrich/aws/privesc/method_54_ecs_create_service.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+cypher: |
+  MATCH (attacker:Principal)-[passrole]->(target)
+  WHERE type(passrole) = 'IAM_PASSROLE'
+  WITH attacker
+  MATCH (attacker)-[create]->(service_resource)
+  WHERE type(create) = 'ECS_CREATESERVICE'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_54_ecs_create_service.yaml]: Fixed. Same scoped-victim pattern applied.

File: pkg/graph/queries/enrich/aws/privesc/method_55_ecs_start_task.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+cypher: |
+  MATCH (attacker:Principal)-[passrole]->(target)
+  WHERE type(passrole) = 'IAM_PASSROLE'
+  WITH attacker
+  MATCH (attacker)-[start]->(service_resource)
+  WHERE type(start) = 'ECS_STARTTASK'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_55_ecs_start_task.yaml]: Fixed. Same scoped-victim pattern applied.

File: pkg/graph/queries/enrich/aws/privesc/method_57_emr_run_job_flow.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+cypher: |
+  MATCH (attacker:Principal)-[passrole]->(target)
+  WHERE type(passrole) = 'IAM_PASSROLE'
+  WITH attacker
+  MATCH (attacker)-[run]->(service_resource)
+  WHERE type(run) = 'ELASTICMAPREDUCE_RUNJOBFLOW'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_57_emr_run_job_flow.yaml]: Fixed. Same scoped-victim pattern applied.

File: pkg/graph/queries/enrich/aws/privesc/method_58_emr_serverless.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+cypher: |
+  MATCH (attacker:Principal)-[passrole]->(target)
+  WHERE type(passrole) = 'IAM_PASSROLE'
+  WITH attacker
+  MATCH (attacker)-[create]->(service_resource)
+  WHERE type(create) = 'EMR-SERVERLESS_CREATEAPPLICATION'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_58_emr_serverless.yaml]: Fixed. Same scoped-victim pattern applied.

File: pkg/graph/queries/enrich/aws/privesc/method_63_imagebuilder_create_pipeline.yaml [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+cypher: |
+  MATCH (attacker:Principal)-[passrole]->(target)
+  WHERE type(passrole) = 'IAM_PASSROLE'
+  WITH attacker
+  MATCH (attacker)-[create]->(service_resource)
+  WHERE type(create) = 'IMAGEBUILDER_CREATEINFRASTRUCTURECONFIGURATION'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_63_imagebuilder_create_pipeline.yaml]: Fixed. Kept `victim` in scope throughout: `MATCH (attacker:Principal)-[passrole]->(victim:Principal) WHERE type(passrole) = 'IAM_PASSROLE' WITH attacker, victim` — victim is carried through all intermediate `WITH` clauses to the final `MERGE`.

File: pkg/graph/queries/enrich/aws/privesc/method_64_kinesis_analytics.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+cypher: |
+  MATCH (attacker:Principal)-[passrole]->(target)
+  WHERE type(passrole) = 'IAM_PASSROLE'
+  WITH attacker
+  MATCH (attacker)-[create]->(service_resource)
+  WHERE type(create) = 'KINESISANALYTICS_CREATEAPPLICATION'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_64_kinesis_analytics.yaml]: Fixed. IAM_PASSROLE target bound as `victim:Principal` and carried through the full pipeline to `MERGE`.

File: pkg/graph/queries/enrich/aws/privesc/method_75_amplify_create_app.yaml [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  WITH attacker, svc
+  MATCH (attacker)-[branch]->(svc)
+  WHERE type(branch) = 'AMPLIFY_CREATEBRANCH'
+  WITH attacker, svc
+  MATCH (attacker)-[job]->(svc)
+  WHERE type(job) = 'AMPLIFY_STARTJOB'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_75_amplify_create_app.yaml]: Fixed. `victim` is now the IAM_PASSROLE target (a `:Principal` node) and is carried through all three intermediate `WITH attacker, victim, svc` steps before the final `WITH DISTINCT attacker, victim MERGE`.

File: pkg/graph/queries/enrich/aws/privesc/method_77_glue_createjob_createtrigger.yaml [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  WITH attacker
+  MATCH (attacker)-[create]->(svc)
+  WHERE type(create) = 'GLUE_CREATEJOB'
+  WITH attacker, svc
+  MATCH (attacker)-[trigger]->(svc)
+  WHERE type(trigger) = 'GLUE_CREATETRIGGER'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_77_glue_createjob_createtrigger.yaml]: Fixed. `WITH attacker` replaced with `WITH attacker, victim` so the passed role is carried through both service action MATCHes.

File: pkg/graph/queries/enrich/aws/privesc/method_79_lambda_passrole_createfunction_addpermission.yaml [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+  WITH attacker
+  MATCH (attacker)-[create]->(svc)
+  WHERE type(create) = 'LAMBDA_CREATEFUNCTION'
+  WITH attacker, svc
+  MATCH (attacker)-[addperm]->(svc)
+  WHERE type(addperm) = 'LAMBDA_ADDPERMISSION'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_79_lambda_passrole_createfunction_addpermission.yaml]: Fixed. Same pattern — `victim` bound at the IAM_PASSROLE MATCH and carried through all intermediate steps.

File: pkg/graph/queries/enrich/aws/privesc/method_81_glue_updatejob_startjobrun.yaml [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  WITH attacker
+  MATCH (attacker)-[update]->(svc)
+  WHERE type(update) = 'GLUE_UPDATEJOB'
+  WITH attacker, svc
+  MATCH (attacker)-[run]->(svc)
+  WHERE type(run) = 'GLUE_STARTJOBRUN'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_81_glue_updatejob_startjobrun.yaml]: Fixed. `WITH attacker` replaced with `WITH attacker, victim` after each service action MATCH.

File: pkg/graph/queries/enrich/aws/privesc/method_86_kinesisanalytics_startapplication.yaml [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+  WITH attacker
+  MATCH (attacker)-[create]->(svc)
+  WHERE type(create) = 'KINESISANALYTICS_CREATEAPPLICATION'
+  WITH attacker, svc
+  MATCH (attacker)-[start]->(svc)
+  WHERE type(start) = 'KINESISANALYTICS_STARTAPPLICATION'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_86_kinesisanalytics_startapplication.yaml]: Fixed. Same scoped-victim pattern applied.

File: pkg/graph/queries/enrich/aws/privesc/method_87_omics_startrun.yaml [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+  WITH attacker
+  MATCH (attacker)-[create]->(svc)
+  WHERE type(create) = 'OMICS_CREATEWORKFLOW'
+  WITH attacker, svc
+  MATCH (attacker)-[run]->(svc)
+  WHERE type(run) = 'OMICS_STARTRUN'
+  WITH attacker
+  MATCH (victim:Principal)
+  WHERE victim.Arn <> attacker.Arn
+  MERGE (attacker)-[pe:CAN_PRIVESC]->(victim)
```
[the remark this answers is not in the document]
Comment (Elgin-Lee) [reply pkg/graph/queries/enrich/aws/privesc/method_87_omics_startrun.yaml]: Fixed. Same scoped-victim pattern applied.

**Review by bernard-yip — APPROVED** [review]
I see some issues, lets fix it in a seperate PR as this one is too large to make any sense
